datumaro/plugins/yolo_format/extractor.py

```diff
@@ -287,11 +315,321 @@ def get_subset(self, name):
-        return cls._find_sources_recursive(path, ".data", "yolo")
+    def name_from_path(cls, path: str) -> str:
+        """
+        Obtains <image name> from the path like [data/]images/<subset>/<image_name>.ext
+
+        <image name> can be <a/b/c/filename>, so it is
+        more involved than just calling "basename()".
+        """
+        path = cls.localize_path(path)
+
+        parts = split_path(path)
+        if 2 < len(parts) and not osp.isabs(path):
+            path = osp.join(*parts[2:])  # pylint: disable=no-value-for-parameter
+        return osp.splitext(path)[0]
+
+    def _iterate_over_image_paths(
+        self, subset_name: str, subset_images_source: Union[str, List[str]]
+    ):
+        if isinstance(subset_images_source, str):
+            if subset_images_source.endswith(YoloPath.SUBSET_LIST_EXT):
+                yield from super()._iterate_over_image_paths(subset_name, subset_images_source)
+            else:
+                path = osp.join(self._path, self.localize_path(subset_images_source))
+                if not osp.isdir(path):
+                    raise InvalidAnnotationError(f"Can't find '{subset_name}' subset image folder")
+                yield from (
+                    osp.relpath(osp.join(root, file), self._path)
+                    for root, dirs, files in os.walk(path)
+                    for file in files
+                    if osp.isfile(osp.join(root, file))
+                )
+        else:
+            yield from subset_images_source
+
+
+class Yolo8SegmentationExtractor(Yolo8Extractor):
+    def _load_segmentation_annotation(
+        self, parts: List[str], image_height: int, image_width: int
+    ) -> Polygon:
+        label_id = self._parse_field(parts[0], int, "Polygon label id")
+        if label_id not in self._categories[AnnotationType.label]:
+            raise UndeclaredLabelError(str(label_id))
+
+        points = [
+            self._parse_field(
+                value, float, f"polygon point {idx // 2} {'x' if idx % 2 == 0 else 'y'}"
+            )
+            for idx, value in enumerate(parts[1:])
+        ]
+        scaled_points = [
+            value * size for value, size in zip(points, cycle((image_width, image_height)))
+        ]
+        return Polygon(scaled_points, label=label_id)
+
+    def _load_one_annotation(
+        self, parts: List[str], image_height: int, image_width: int
+    ) -> Annotation:
+        if len(parts) > 5 and len(parts) % 2 == 1:
+            return self._load_segmentation_annotation(parts, image_height, image_width)
+        raise InvalidAnnotationError(
+            f"Unexpected field count {len(parts)} in the polygon description. "
+            "Expected odd number > 5 of fields for segment annotation (label, x1, y1, x2, y2, x3, y3, ...)"
+        )
+
+
+class Yolo8OrientedBoxesExtractor(Yolo8Extractor):
+    @staticmethod
+    def _check_is_rectangle(p1, p2, p3, p4):
+        p12_angle = math.atan2(p2[0] - p1[0], p2[1] - p1[1])
```

Comment:
probably it is possible to do that without math library.
1. My way is to check that opposite sides are equal and after that check a^2 + b^2 == c^2
2. On the internet I have found another way. Find the center of mass and calculate distance till all points from the center mass. They should be equal.

Because the check will be run multiple times, I would recommend to choose the optimal implementation. Can you compare them?

Reply:
Compared my implementation
```python
def _check_is_rectangle_atan2(p1, p2, p3, p4):
    p12_angle = math.atan2(p2[0] - p1[0], p2[1] - p1[1])
    p23_angle = math.atan2(p3[0] - p2[0], p3[1] - p2[1])
    p43_angle = math.atan2(p3[0] - p4[0], p3[1] - p4[1])
    p14_angle = math.atan2(p4[0] - p1[0], p4[1] - p1[1])

    if abs(p12_angle - p43_angle) > 0.001 or abs(p23_angle - p14_angle) > 0.001:
        return False
    return True
```
with your second option. I have not tried your option 1, because it will have more pows and more sqrts, therefore should be slower than option 2.
```python
def _check_is_rectangle_2(p1, p2, p3, p4):
    center_x = (p1[0] + p2[0] + p3[0] + p4[0]) / 4
    center_y = (p1[1] + p2[1] + p3[1] + p4[1]) / 4

    distances_squared = [
        (x - center_x) ** 2 + (y - center_y) ** 2
        for x, y in (p1, p2, p3, p4)
    ]

    if math.sqrt(max(distances_squared)) - math.sqrt(min(distances_squared)) > 0.1:
        return False
    return True
```
Results:
```python
%%timeit
_check_is_rectangle_atan2(f_p1, f_p2, f_p3, f_p4)
```
475 ns ± 6.64 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```python
%%timeit
_check_is_rectangle_2(f_p1, f_p2, f_p3, f_p4)
```
1.52 μs ± 9.1 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

I tried to do it without a list, did not hrlp much. 
I have no idea why atan2 implementation is faster

Reply:
The best I managed is that:
```python
def _check_is_rectangle_vector(p1, p2, p3, p4):
    # check vectors p1->p2 and p3->p4 sum to 0
    if abs(p2[0] - p1[0] + p4[0] - p3[0]) > 0.001 or abs(p2[1] - p1[1] + p4[1] - p3[1]) > 0.001:
        return False
    # check vectors p2->p3 and p4->p1 sum to 0
    if abs(p3[0] - p2[0] + p1[0] - p4[0]) > 0.001 or abs(p3[1] - p2[1] + p1[1] - p4[1]) > 0.001:
        return False
    return True
```
338 ns ± 5.1 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each) 

In case  if points do not form a rectangle it works even faster. But this code also is less understandable, so I am not sure it would be justifiable to use it instead of atan2 version